### DISCUSSION TITLE
How to set admin role to have access in viewing and deleting log?

### DISCUSSION THREAD

**Discussion by anonymouse703** [opcodesio/log-viewer#91] [Q&A] [answered]
Base on the code you can set it by email for the user to view the log but how about by role (admin role)?

  ```
LogViewer::auth(function ($request) {
        return $request->user()
            && in_array($request->user()->email, [
                // 'john@example.com',
            ]);
    });
```
and also on the download and delete?

download
 ```
$this->registerPolicies();
 
    Gate::define('downloadLogFile', function (?User $user, LogFile $file) {
        // return true if the user is allowed to download the specific log file.
    });
```
and delete

```
 $this->registerPolicies();
 
    Gate::define('deleteLogFile', function (?User $user, LogFile $file) {
        // return true if the user is allowed to delete the specific log file.
    });
    ```

**Comment by arukompas** [comment, marked as answer]
Within the callback you can execute any code you want. The above is just an example.

So use however you normally use roles. 

For example, if your roles package or implementation has `hasRole('admin')` method, then you would simply use that in the callback:

```php
LogViewer::auth(fn ($request) => $request->user()->hasRole('admin'));
```

From the callback you just need to return `true` to allow access, and `false` to forbid access.

**Reply by anonymouse703** [reply]
Thanks sir, how about on the download and delete? how to  turn true and false? I mean what is the syntax on it?

**Reply by arukompas** [reply]
@anonymouse703 use the examples in the docs, but replace the comments with actual PHP logic to return `true` to allow, or `false` to forbid.

For example:
```diff
    Gate::define('downloadLogFile', function (?User $user, LogFile $file) {
-        // return true if the user is allowed to download the specific log file.
+        return $user->hasRole('admin');
    });
```

**Reply by anonymouse703** [reply]
@arukompas OKay thanks sir. I was just confuse with the second parameter

**Reply by arukompas** [reply]
@anonymouse703 - the second parameter is the log file they're trying to download or delete. You don't have to use that, but sometimes it can be useful. For example, if you want to allow downloading all files **except** `errors.log`:

```php
    Gate::define('downloadLogFile', function (?User $user, LogFile $file) {
        return $user->hasRole('admin')
            && $file->name !== 'errors.log';
    });
```
